Incorporate gpu support from upstream

JaimeFrey opened the pull request:
This is a modified form the the GPU support added to the blahp in the prelz codebase in commit 444faae.

JaimeFrey committed:
Add support for GPUs. #7864

Add new blahp job attributes MICNumber, GPUNumber, GPUMode, and
GPUModel. Add to the batch submit scripts where supported.

This is a modified form of a commit made by Paolo Andreetto
<paolo.andreetto@pd.infn.it>.

brianhlin commented:
@edquist can you add a review of the bash bits? I'm contacting the Operations team about validating this with real job submission.

edquist commented on src/scripts/blah_common_submit_functions.sh:
```diff
@@ -314,6 +314,10 @@ function bls_parse_submit_options ()
       z) bls_opt_wholenodes="$OPTARG";;
       h) bls_opt_hostnumber="$OPTARG";;
+      g) bls_opt_gpunumber="$OPTARG";;
+      m) bls_opt_gpumode="$OPTARG";;
```
This is an issue, there is already a `-m` option (`bls_opt_req_mem`)

JaimeFrey replied:
I will change -m to -G.

edquist commented on src/scripts/lsf_submit.sh:
```diff
@@ -142,6 +142,31 @@ else
 
+# --- Begin of GPU directives
+
+if [ ! -z $bls_opt_gpunumber ] && [ $bls_opt_gpunumber -gt 0 ] ; then
```
is there a default value for these options? (`$bls_opt_gpunumber` and  `$bls_opt_gpunumber`)

if not, this looks to me like it will blow chunks if these vars are empty/unset

JaimeFrey replied:
I pasted that code block into a bare bash script and it happily treated the conditional as false.

edquist replied:
To me it's one of those "i do not think that word means what you think it means" moments.

You are right, it works if `bls_opt_gpunumber` is unset, though only incidentally.

`[ ! -z $bls_opt_gpunumber ]` becomes `[ ! -z ]` which means, 'test that the length of the string "-z" is not non-zero', which is false; thus `[ $bls_opt_gpunumber -gt 0 ]` (which would blow chunks) is not evaluated at all due to the `&&` short circuit.

So, incidentally, the behavior is logically equivalent to what was intended.

So it's one of those "if you break it enough, it just might work" kind of things.

Anyway, i only give my approval in the "obfuscated code contest" sense.

edquist commented on src/scripts/lsf_submit.sh:
```diff
@@ -142,6 +142,31 @@ else
+  if [ ! -z $bls_opt_gpumodel ] ; then
+       model="gpu_model0=='$bls_opt_gpumodel'"
+       if [ $bls_opt_gpunumber -gt 1 ] ; then
+        for i in $(seq 1 $((bls_opt_gpunumber-1)))
```
the indentation in this block is intolerable.

Also, does the author know about this `for` loop syntax?

```suggestion
        for ((i = 1; i < bls_opt_gpunumber; i++))
```

tannenba replied:
In Jaime's defense, both the for loop syntax and the indenting originated upstream... is it worth converting things like for-loop syntax to something different (albeit perhaps better) at the cost of making incorporating future upstream patches that much more difficult?

JaimeFrey replied:
I will further plead that the upstream scope also has the duplicate 'm' argument. We could drop the GPUMode attribute (and probably MICNumber as well).

edquist replied:
Ok - leave the indentation and for-loop like it appears in prelz's repo, if you like.

edquist commented on src/scripts/pbs_submit.sh:
```diff
@@ -154,27 +154,37 @@ bls_set_up_local_and_extra_args
 
-# Extended support for MPI attributes
+# Extended support for MPI attributes (+GPU for pbs_sched)
+pmaui=$(/usr/bin/pgrep maui)
```
this seems like a terrible idea

edquist replied:
I mean, a more exact process name match, at least...

edquist replied:
Anyway, just seems real easy for any old user to fool this script by running a program named maui.  If that's really the way we have to do it, at least consider adding the `-x` option to pgrep to get an exact process name match, although even that still feels pretty sketchy to me.

JaimeFrey replied:
I'll add a pbs_maui parameter to the blah config file (like the existing pbs_pro parameter), defaulting to "no".

edquist commented on src/scripts/slurm_submit.sh:
```diff
@@ -85,6 +85,14 @@ if [[ $bls_opt_mpinodes -gt 1 ]] ; then
 fi
 
+# add GPU support
+[ -z "$bls_opt_gpunumber" ] || [ -z "$bls_opt_gpumodel" ] || echo "#SBATCH --gres=gpu:${bls_opt_gpumodel}:${bls_opt_gpunumber}" >> $bls_tmp_file
+[ -z "$bls_opt_gpunumber" ] || [ ! -z "$bls_opt_gpumodel" ] || echo "#SBATCH --gres=gpu:${bls_opt_gpunumber}" >> $bls_tmp_file
+[ ! -z "$bls_opt_gpunumber" ] || [ -z "$bls_opt_gpumodel" ] || echo "#SBATCH --gres=gpu:${bls_opt_gpumodel}:1" >> $bls_tmp_file
```
Just a guess, but i suspect the logic might be wrong here, with all the extra negations.

I'd bet a beer, actually.

I'd suggest using `if` blocks or `&&` for these conditions -- it's kind of easy to get wrong otherwise.  And it's not easy to follow, reading it, either.

edquist replied:
Ah, after closer inspection, maybe it is correct.  But it is a bit painful to try to make sense of it.

JaimeFrey replied:
Agreed on the confusing syntax. Barring bugs, I'd prefer to leave it alone to avoid divergence from upstream.

edquist replied:
Ok, given that you got it from somewhere else.

edquist reviewed (changes requested):
The most serious issue is the `-m` option is already used... Have to pick a different option name, i think.

edquist commented:
Sorry Jaime if i came off harsh in my review - i don't blame you for any of it.  All of my critiques are directed at the code, not at whoever happened to write or transcribe it.   There is a certain style the blah is famous for, i guess.  But the `-m` thing is actually broken, and i'd be remiss in not saying so in a code review.

bbockelm commented:
The `grep` through all processes on the system to see if it is running Maui seems pretty dicey... especially if it occurs for every single job submit.

How often does the PBS variant change during the lifetime of the CE installation?

JaimeFrey commented:
No worries, Carl. I did a fairly uncritical merge of the changes, hence wanting a code review.

BrianB, that does look like an unnecessary expense on every submit.

JaimeFrey committed:
Address code review issues for GPU support. #7864

Add pbs_maui config variable to avoid a sketchy pgrep call on every
submit for PBS.
Change command-line switch for GPUMode from -m (already used) to -G.

JaimeFrey commented:
I've made changes based on Carl's review.

edquist reviewed (approved):
Thanks Jaime for addressing especially the duplicate -m option.

I'd imagine ideally we'd improve the other parts too and suggest that prelz take the improvements, but that doesn't have to be here.
